tests/Credentials/InstanceProfileProviderTest.php

```diff
@@ -1349,4 +1349,134 @@ private function fetchMockedCredentialsAndAlwaysExpectAToken($config=[]) {
             return false;
         }
     }
+
+    /**
+     * @dataProvider endpointModeCases
+     */
+    public function testEndpointMode($clientConfig, $env, $config, $expected)
```

Comment:
Let's modularize the tests here.  We should have separate tests for both resolving/validating provided endpoints, as well as using the provided endpoint mode